### PR TITLE
Model: add `RequirementCategory` Model

### PR DESCRIPTION
**Pull request by chyeo**
The `RequirementCategory` Model is not yet implemented in the codes. 

Let's add the `RequirementCategory` Model into the codes in a step by step manner:
- create the `RequirementCategory` class
- create the `UniqueRequirementCategoryList` 
- adding `RequirementCategory` exceptions
- create the `RequirementCategoryModel` 
- update the `RequirementCategoryModel` to handle the `RequirementCategory` class
- update the 'UserPref' to prepare for the storage of the `RequirementCategory` class
- creating sample data for the `RequirementCategory` class
- creating a `RequirementCategoryList` storage to store `Requirement`
- creating Json files to manipulate the storage data
- update the `MainApp` to ensure that the storage is working with the `RequirementCategoryModel` and `RequirementCategory` class

The following are the commits for easy referencing:
* [ 1/21] [Requirement: add `Requirement` Model](https://github.com/CS2113-AY1819S2-T09-1/main/pull/43/commits/9f5609cfbbec54a5d46ce2d293422ce4adfc9ba2)
* [ 2/21] [Requirement: add UniqueRequirementList](https://github.com/CS2113-AY1819S2-T09-1/main/pull/43/commits/fbf205539f9ff0503aeb8a19c6068349d2e53098)
* [ 3/21] [Requirement: add ModuleList](https://github.com/CS2113-AY1819S2-T09-1/main/pull/43/commits/9c16b81a4d84d20afc68b42bbd2d62bfca2d1354)
* [ 4/21] [Requirement: add Requirement exceptions](https://github.com/CS2113-AY1819S2-T09-1/main/pull/43/commits/5d4e514b841eda61fd7c152569d8c74aeb9b68ba)
* [ 5/21] [Model: add RequirementModel](https://github.com/CS2113-AY1819S2-T09-1/main/pull/43/commits/584f1e5a8a7a89f58503d601548730d03ebeb7ab)
* [ 6/21] [Model: add RequirementList](https://github.com/CS2113-AY1819S2-T09-1/main/pull/43/commits/7c2945c03f4a7e363c72995d538cf15d1b174b1c)
* [ 7/21] [Model: add ReadOnlyRequirementList](https://github.com/CS2113-AY1819S2-T09-1/main/pull/43/commits/1da3f02e0cc2e34531193de79869e7710eefa7b0)
* [ 8/21] [Model: add VersionedRequirementList](https://github.com/CS2113-AY1819S2-T09-1/main/pull/43/commits/9515c1c7358bc8544e179084e736056b164b4a6c)
* [ 9/21] [Model: add ModelManager methods](https://github.com/CS2113-AY1819S2-T09-1/main/pull/43/commits/c79f0d1075888ea2d9c3316cc925abb7935e44ee)
* [10/21] [Model: update to handle Requirement](https://github.com/CS2113-AY1819S2-T09-1/main/pull/43/commits/63924a9c12486a42c0602fcc0e1d21351905aec4)
* [11/21] [Model: update UserPref to prep Requirement storage](https://github.com/CS2113-AY1819S2-T09-1/main/pull/43/commits/fe2f5f073e787713c841a15b57c8f789082e33e5)
* [12/21] [Model: add sample data generator for Requirement](https://github.com/CS2113-AY1819S2-T09-1/main/pull/43/commits/8d37eb9695218cd0ff721527619e7d9a392941cd)
* [13/21] [Storage: add RequirementList storage](https://github.com/CS2113-AY1819S2-T09-1/main/pull/43/commits/8b0fc531167de6a3276e01cb247b92fc2b639926)
* [14/21] [Storage: add Json files to handle RequirementList](https://github.com/CS2113-AY1819S2-T09-1/main/pull/43/commits/79abf7d40332733058dcc41bd93a77935af628a9)
* [15/21] [Storage: update to handle RequirementList](https://github.com/CS2113-AY1819S2-T09-1/main/pull/43/commits/3d05c6893f7f7518395efd93ddff135d99f84b0e)
* [16/21] [MainApp: update to handle storage for RequirementList](https://github.com/CS2113-AY1819S2-T09-1/main/pull/43/commits/4ed3b4f3d061f1ba88e7d19323962fdd6e5c95df)
* [17/21] [Checkstyle errors correction](https://github.com/CS2113-AY1819S2-T09-1/main/pull/43/commits/ae8bbe6538857a40af41dafc39cd11414f14b68a)
* [18/21] [Test: update to include RequirementList](https://github.com/CS2113-AY1819S2-T09-1/main/pull/43/commits/2f71e827d6ab04122206dd111649a494be39776f)
* [19/21] [Model: refractor 'Requirement' to 'RequirementCategory'](https://github.com/CS2113-AY1819S2-T09-1/main/pull/43/commits/3a753bb1b97baa49435c680ed3527d88cd13929c)
* [20/21] [Fix Checkstyle Errors](https://github.com/CS2113-AY1819S2-T09-1/main/pull/43/commits/b545d8549ae6e32d732cdd693ee11b4f71acb52c)
* [21/21] [Model: update SampleRequirementCategoryUtil](https://github.com/CS2113-AY1819S2-T09-1/main/pull/43/commits/5d1b3086d3be17a1d781b880905e6e46a6a5c2ba)

### CODE REVIEW

File: src/main/java/seedu/address/model/ModelManager.java [hunk 1]
```diff
@@ -25,40 +27,46 @@
     private static final Logger logger = LogsCenter.getLogger(ModelManager.class);
 
     private final VersionedAddressBook versionedAddressBook;
+    private final VersionedRequirementList versionedRequirementList;
```
Comment (Creastery) [inline src/main/java/seedu/address/model/ModelManager.java]: As discussed, I am accepting this approach for v1.1 milestone.
However, let's aim to combine this with `VersionedAddressBook` by v1.3 milestone.

File: src/main/java/seedu/address/model/requirement/Requirement.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+import java.util.Set;
+
+import seedu.address.model.module.Credits;
+import seedu.address.model.module.Name;
+
+/**
+ * Represents a Requirement in the requirement list.
+ * Guarantees: details are present and not null, field values are validated, immutable.
+ */
+public class Requirement {
```
Comment (Creastery) [inline src/main/java/seedu/address/model/requirement/Requirement.java]: `Requirement` -> `RequirementCategory`

File: src/main/java/seedu/address/model/requirement/Requirement.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+    // Identity fields
+    private final Name name;
+
+    // Data fields
+    private final Credits credits;
+    private final Set<ModuleList> moduleList = new HashSet<>();
```
Comment (Creastery) [inline src/main/java/seedu/address/model/requirement/Requirement.java]: `Set<ModuleList> moduleList` should be `Set<Code> codes`.
We shouldn't be duplicating the list of `Module`.
Rather, we should maintain a list of `Code`, and lookup `ModuleList` to find the respective module.

File: src/main/java/seedu/address/model/requirement/UniqueRequirementList.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+ * adding and updating of requirement uses Requirement#isSameRequirement(Requirement) for equality so as to
+ * ensure that the requirement being added or updated is unique in terms of identity in the UniqueRequirementList.
+ * However, the removal of a requirement uses Requirement#equals(Object) so as to ensure that the requirement with exactly
+ * the same fields will be removed.
+ * <p>
+ * Supports a minimal set of list operations.
+ *
+ * @see Requirement#isSameRequirement(Requirement)
+ */
+public class UniqueRequirementList implements Iterable<Requirement> {
```
Comment (Creastery) [inline src/main/java/seedu/address/model/requirement/UniqueRequirementList.java]: `UniqueRequirementList` -> `UniqueRequirementCategoryList`

File: src/main/java/seedu/address/model/requirement/ModuleList.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
+package seedu.address.model.requirement;
+
+import static java.util.Objects.requireNonNull;
+import static seedu.address.commons.util.AppUtil.checkArgument;
+
+/**
+ * Represents a ModuleList in the address book.
+ * Guarantees: immutable; name is valid as declared in {@link #isValidModuleName(String)}
+ */
+public class ModuleList {
```
Comment (Creastery) [inline src/main/java/seedu/address/model/requirement/ModuleList.java]: Remove this `ModuleList` class.

File: src/main/java/seedu/address/model/RequirementModel.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+import java.util.function.Predicate;
+
+import javafx.collections.ObservableList;
+import seedu.address.model.requirement.Requirement;
+
+/**
+ * The API of the Model component.
+ */
+public interface RequirementModel {
```
Comment (Creastery) [inline src/main/java/seedu/address/model/RequirementModel.java]: `RequirementModel` -> `RequirementCategoryModel`

File: src/main/java/seedu/address/model/RequirementList.java [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+import seedu.address.commons.util.InvalidationListenerManager;
+import seedu.address.model.requirement.Requirement;
+import seedu.address.model.requirement.UniqueRequirementList;
+
+/**
+ * Wraps all data at the requirement-list level
+ * Duplicates are not allowed (by .isSameRequirement comparison)
+ */
+
+public class RequirementList implements ReadOnlyRequirementList {
```
Comment (Creastery) [inline src/main/java/seedu/address/model/RequirementList.java]: `RequirementList` -> `RequirementCategoryList`

File: src/main/java/seedu/address/model/ReadOnlyRequirementList.java [hunk 8]
```diff
@@ -0,0 +1,18 @@
+package seedu.address.model;
+
+import javafx.beans.Observable;
+import javafx.collections.ObservableList;
+import seedu.address.model.requirement.Requirement;
+
+/**
+ * Unmodifiable view of planner list
+ */
+public interface ReadOnlyRequirementList extends Observable {
```
Comment (Creastery) [inline src/main/java/seedu/address/model/ReadOnlyRequirementList.java]: `ReadOnlyRequirementList` -> `ReadOnlyRequirementCategoryList`

File: src/main/java/seedu/address/model/VersionedRequirementList.java [hunk 9]
```diff
@@ -0,0 +1,110 @@
+package seedu.address.model;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * {@code RequirementList} that keeps track of its own history.
+ */
+public class VersionedRequirementList extends RequirementList {
```
Comment (Creastery) [inline src/main/java/seedu/address/model/VersionedRequirementList.java]: `VersionedRequirementList` -> `VersionedRequirementCategoryList`

File: src/main/java/seedu/address/model/util/SampleRequirementUtil.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+import seedu.address.model.module.Credits;
+import seedu.address.model.module.Name;
+import seedu.address.model.requirement.Requirement;
+import seedu.address.model.requirement.ModuleList;
+
+public class SampleRequirementUtil {
+    public static Requirement[] getSampleRequirements() {
+        return new Requirement[] {
+                new Requirement(new Name("Computing Foundation"), new Credits("20"),
+                        getModuleSet("CS2100")),
```
Comment (Creastery) [inline src/main/java/seedu/address/model/util/SampleRequirementUtil.java]: Should use `getModuleSet()` (empty arguments).

File: src/main/java/seedu/address/storage/RequirementListStorage.java [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+import java.nio.file.Path;
+import java.util.Optional;
+
+import seedu.address.commons.exceptions.DataConversionException;
+import seedu.address.model.ReadOnlyRequirementList;
+
+    /**
+     * Represents a storage for {@link seedu.address.model.RequirementList}.
+     */
+    public interface RequirementListStorage {
```
Comment (Creastery) [inline src/main/java/seedu/address/storage/RequirementListStorage.java]: `RequirementListStorage` -> `RequirementCategoryListStorage`

File: src/main/java/seedu/address/storage/JsonAdaptedModuleList.java [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonValue;
+
+import seedu.address.commons.exceptions.IllegalValueException;
+import seedu.address.model.requirement.ModuleList;
+
+/**
+ * Jackson-friendly version of {@link ModuleList}.
+ */
+class JsonAdaptedModuleList {
```
Comment (Creastery) [inline src/main/java/seedu/address/storage/JsonAdaptedModuleList.java]: `JsonAdaptedModuleList` can be deleted.

File: src/main/java/seedu/address/storage/JsonSerializableRequirementList.java [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+import seedu.address.commons.exceptions.IllegalValueException;
+import seedu.address.model.RequirementList;
+import seedu.address.model.ReadOnlyRequirementList;
+import seedu.address.model.requirement.Requirement;
+
+    /**
+     * An Immutable RequirementList that is serializable to JSON format.
+     */
+    @JsonRootName(value = "requirementList")
+    public class JsonSerializableRequirementList {
```
Comment (Creastery) [inline src/main/java/seedu/address/storage/JsonSerializableRequirementList.java]: `JsonSerializableRequirementList` -> `JsonSerializableRequirementCategoryList`

File: src/main/java/seedu/address/storage/JsonRequirementListStorage.java [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+import seedu.address.commons.exceptions.DataConversionException;
+import seedu.address.commons.exceptions.IllegalValueException;
+import seedu.address.commons.util.FileUtil;
+import seedu.address.commons.util.JsonUtil;
+import seedu.address.model.ReadOnlyRequirementList;
+
+    /**
+     * A class to access RequirementList data stored as a json file on the hard disk.
+     */
+    public class JsonRequirementListStorage implements RequirementListStorage {
```
Comment (Creastery) [inline src/main/java/seedu/address/storage/JsonRequirementListStorage.java]: `JsonRequirementListStorage` -> `JsonRequirementCategoryListStorage`

File: src/main/java/seedu/address/storage/JsonAdaptedRequirementList.java [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+import seedu.address.commons.exceptions.IllegalValueException;
+import seedu.address.model.requirement.Requirement;
+import seedu.address.model.requirement.ModuleList;
+import seedu.address.model.module.Credits;
+import seedu.address.model.module.Name;
+
+/**
+ * Jackson-friendly version of {@link Requirement}.
+ */
+public class JsonAdaptedRequirementList {
```
Comment (Creastery) [inline src/main/java/seedu/address/storage/JsonAdaptedRequirementList.java]: `JsonAdaptedRequirementList` -> `JsonAdaptedRequirementCategoryList`

File: src/main/java/seedu/address/storage/JsonRequirementListStorage.java [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+        try {
+            return Optional.of(jsonRequirementList.get().toModelType());
+        } catch (IllegalValueException ive) {
+            logger.info("Illegal values found in " + filePath + ": " + ive.getMessage());
+            throw new DataConversionException(ive);
+        }
+    }
+
+    @Override
+    public void saveRequirementList(ReadOnlyRequirementList requirementList) throws IOException {
```
Comment (Creastery) [inline src/main/java/seedu/address/storage/JsonRequirementListStorage.java]: This function is never invoked during the application lifespan.
I think you need to trigger it explicitly in `LogicManager#execute(...)`.

File: src/main/java/seedu/address/model/ModelManager.java (hunk 1 above)
Comment (chyeo) [reply src/main/java/seedu/address/model/ModelManager.java]: Roger, will work with @truegitnovice to combine into `VersionedAddressBook`

File: src/main/java/seedu/address/model/requirement/Requirement.java (hunk 3 above)
Comment (chyeo) [reply src/main/java/seedu/address/model/requirement/Requirement.java]: Roger, will make the necessary changes

File: src/main/java/seedu/address/model/requirement/ModuleList.java (hunk 5 above)
Comment (chyeo) [reply src/main/java/seedu/address/model/requirement/ModuleList.java]: Noted, will remove `ModuleList` class

File: src/main/java/seedu/address/storage/JsonAdaptedModuleList.java (hunk 12 above)
Comment (chyeo) [reply src/main/java/seedu/address/storage/JsonAdaptedModuleList.java]: Noted, will delete `JsonAdaptedModuleList`

File: src/main/java/seedu/address/storage/JsonRequirementListStorage.java (hunk 16 above)
Comment (chyeo) [reply src/main/java/seedu/address/storage/JsonRequirementListStorage.java]: Roger, will make the amendments to call the relevant methods

**Review by Creastery — APPROVED** [review]
This PR became way too large. The original intent conveyed in the PR was to add the `RequirementCategory` Model, and became everything else associated with it to enable this model to function.

Please try to avoid large PRs to implement multiple functionalities at one go next time.

Nevertheless, given the effort put into doing this, and after verifying that it works (through force-saving the files and editing the files), I am inclined to approve and merge these changes.

We can iron out the remaining issues in the next milestone.
Thanks for the effort!

**Pull request merged**